LocationManager/INTULocationManager/INTULocationManager.m

```diff
@@ -692,6 +758,110 @@ - (INTULocationAccuracy)achievedAccuracyForLocation:(CLLocation *)location
+ */
+- (void)addHeadingRequest:(INTUHeadingRequest *)headingRequest
+{
+    NSAssert(headingRequest, @"Must pass in a non-nil heading request.");
+
+    // If heading services are not available, just return
+    if ([INTULocationManager headingServicesState] == INTUHeadingServicesStateUnavailable) {
+        if (headingRequest.block) {
+            headingRequest.block(nil, INTUHeadingStatusUnavailable);
+        }
+        INTULMLog(@"Heading Request (ID %ld) NOT added since device heading is unavailable.", (long)headingRequest.requestID);
+        return;
+    }
+
+    __INTU_GENERICS(NSMutableArray, INTUHeadingRequest *) *newHeadingRequests = [NSMutableArray arrayWithArray:self.headingRequests];
+    [newHeadingRequests addObject:headingRequest];
+    self.headingRequests = newHeadingRequests;
+    INTULMLog(@"Heading Request added with ID: %ld", (long)headingRequest.requestID);
+
+    if (!self.isUpdatingHeading) {
+        self.isUpdatingHeading = YES;
+
+        [self.locationManager startUpdatingHeading];
+    }
+}
+
+/**
+ Removes a given heading request from the array of requests and stops heading updates if needed.
+ */
+- (void)removeHeadingRequest:(INTUHeadingRequest *)headingRequest
+{
+    __INTU_GENERICS(NSMutableArray, INTUHeadingRequest *) *newHeadingRequests = [NSMutableArray arrayWithArray:self.headingRequests];
+    [newHeadingRequests removeObject:headingRequest];
+    self.headingRequests = newHeadingRequests;
+
+    [self stopUpdatingHeadingIfPossible];
+}
+
+/**
+ Checks to see if there are any outstanding headingRequests, and if there are none, informs CLLocationManager to stop sending
+ heading updates. This is done as soon as heading updates are no longer needed in order to conserve the device's battery.
+ */
+- (void)stopUpdatingHeadingIfPossible
+{
+    if (self.headingRequests.count == 0) {
+        [self.locationManager stopUpdatingHeading];
+        if (self.isUpdatingHeading) {
+            INTULMLog(@"Location services heading updates have stopped.");
+        }
+        self.isUpdatingHeading = NO;
+    }
+}
+
+/**
+ Iterates over the array of active heading requests and processes each
+ */
+- (void)processRecurringHeadingRequests
+{
+    for (INTUHeadingRequest *headingRequest in self.headingRequests) {
+        [self processRecurringHeadingRequest:headingRequest];
+    }
+}
+
+/**
+ Handles calling a recurring heading request's block with the current heading.
+ */
+- (void)processRecurringHeadingRequest:(INTUHeadingRequest *)headingRequest
+{
+    NSAssert(headingRequest.isRecurring, @"This method should only be called for recurring heading requests.");
+
+    INTUHeadingStatus status = [self statusForHeadingRequest:headingRequest];
+
+    // Check if the request had a fatal error and should be canceled
+    if (status == INTUHeadingStatusUnavailable) {
+        // No need for dispatch_async when calling this block, since this method is only called from a CLLocationManager callback
```

Comment:
Let's go ahead and dispatch_async here too, like the recent change I made for the corresponding location updates code path. This allows us to safely process the existing requests at any time.